cmd/uniq/uniq.go

```diff
@@ -0,0 +1,99 @@
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+)
+
+type options struct {
+	printDuplicates bool
+	printEmptyLines bool
+	printLineNumber bool
+}
+
+func parseArgs(args []string) options {
+	var opts options
+
+	for _, opt := range args[1:] {
+		switch opt {
+		case "-dup":
+			opts.printDuplicates = true
+		case "-empty":
+			opts.printEmptyLines = true
+		case "-num":
+			opts.printLineNumber = true
+		default:
+			fmt.Println("Usage:")
+			fmt.Println("uniq [-dup | -empty | -num]")
+			os.Exit(1)
+		}
+	}
+	return opts
+}
+
+func uniq(input *os.File, opts options) {
+	inputMap := make(map[string]*string)
+	lineNumsMap := make(map[*string][]int)
+	lineMap := make(map[int]*string)
+
+	reader := bufio.NewReader(input)
+	i := 0
+	for {
+		i++
+		lineBytes, err := reader.ReadBytes(byte('\n'))
+		if err == io.EOF {
+			break
+		}
+
+		line := string(lineBytes)
+		var lineP *string
+		if val, has := inputMap[line]; has {
+			lineP = val
+		} else {
+			lineP = &line
+		}
+		lineCount := len(lineNumsMap[lineP])
+
+		if lineCount == 0 || (opts.printEmptyLines && *lineP == "\n") {
+			lineMap[i] = lineP
+		} else if lineCount == 1 && !opts.printDuplicates {
+			continue
+		}
+		inputMap[line] = lineP
+		lineNumsMap[lineP] = append(lineNumsMap[lineP], i)
+	}
+
+	lineMapLen := len(lineMap)
+	printed := 0
+	for ln := 1; ln < i; ln++ {
+		if printed == lineMapLen {
+			break
+		}
+		if _, has := lineMap[ln]; has == false {
+			continue
+		}
+		lineP := lineMap[ln]
+		lineNumbers := lineNumsMap[lineP]
+		numbersCount := len(lineNumbers)
```

Comment:
It will, but len do not count because strings are immutable. Golang string store its length internally, then len just access it.